tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx

```diff
@@ -377,142 +377,116 @@ public:
-      }
-      auto x0sIt = std::begin(x0s);
-      const auto x0sEnd = std::end(x0s);
-      auto x1sIt = std::begin(x1s);
-      auto x2sIt = std::begin(x2s);
-      for (; x0sIt != x0sEnd; x0sIt++, x1sIt++, x2sIt++) {
-         thisSlotH->Fill(*x0sIt, *x1sIt, *x2sIt); // TODO: Can be optimised in case T == vector<double>
-      }
+      return 1;
    }
 
-   template <
-      typename X0, typename X1, typename W,
-      std::enable_if_t<IsDataContainer<X0>::value && IsDataContainer<X1>::value && !IsDataContainer<W>::value, int> = 0>
-   void Exec(unsigned int slot, const X0 &x0s, const X1 &x1s, const W w)
+   // return container size
+   template <typename T, typename std::enable_if<!std::is_convertible<T, double>::value, int>::type = 0>
+   std::size_t GetSize(const T &val)
    {
-      auto thisSlotH = fObjects[slot];
-      if (x0s.size() != x1s.size()) {
-         throw std::runtime_error("Cannot fill histogram with values in containers of different sizes.");
-      }
-      auto x0sIt = std::begin(x0s);
-      const auto x0sEnd = std::end(x0s);
-      auto x1sIt = std::begin(x1s);
-      for (; x0sIt != x0sEnd; x0sIt++, x1sIt++) {
-         thisSlotH->Fill(*x0sIt, *x1sIt, w); // TODO: Can be optimised in case T == vector<double>
-      }
+#if __cplusplus >= 201703L
+      return std::size(val);
+#else
+      return val.size();
+#endif
    }
 
-   template <typename X0, typename X1, typename X2, typename X3,
-             std::enable_if_t<IsDataContainer<X0>::value && IsDataContainer<X1>::value && IsDataContainer<X2>::value &&
-                                 IsDataContainer<X3>::value,
-                              int> = 0>
-   void Exec(unsigned int slot, const X0 &x0s, const X1 &x1s, const X2 &x2s, const X3 &x3s)
+   template <std::size_t ColIdx, typename End_t, typename... Its>
+   void ExecLoop(unsigned int slot, End_t end, Its... its)
    {
-      auto thisSlotH = fObjects[slot];
-      if (!(x0s.size() == x1s.size() && x1s.size() == x2s.size() && x1s.size() == x3s.size())) {
-         throw std::runtime_error("Cannot fill histogram with values in containers of different sizes.");
-      }
-      auto x0sIt = std::begin(x0s);
-      const auto x0sEnd = std::end(x0s);
-      auto x1sIt = std::begin(x1s);
-      auto x2sIt = std::begin(x2s);
-      auto x3sIt = std::begin(x3s);
-      for (; x0sIt != x0sEnd; x0sIt++, x1sIt++, x2sIt++, x3sIt++) {
-         thisSlotH->Fill(*x0sIt, *x1sIt, *x2sIt, *x3sIt); // TODO: Can be optimised in case T == vector<double>
+      auto *thisSlotH = fObjects[slot];
+      // loop increments all of the iterators while leaving scalars unmodified
+      // TODO this could be simplified with fold expressions or std::apply in C++17
+      auto nop = [](auto &&...) {};
+      for (; GetNthElement<ColIdx>(its...) != end; nop(Increment(its)...)) {
+         // dereference iterators, but pass scalars through unmodified
+         thisSlotH->Fill(Deref(its)...);
```

Comment:
If I'm not mistaken we could get rid of the Deref helper function and always use `*its` if MakeBegin returned a pointer instead of a reference for scalars.

Reply:
Yes, good point

Reply:
Implementation is simplified and should also address https://github.com/root-project/root/issues/9428

Reply:
amazing, thanks a lot!